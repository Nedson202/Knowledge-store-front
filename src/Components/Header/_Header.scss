.navbar {
  justify-content: space-between !important;

  .container {
    margin: 0;
    max-width: 100%;
  }

  &-brand {
    margin-left: 0.5rem;
  }

  .dropdown-menu {
    left: unset !important;
    right: 10%;
  }
}

@media (min-width: 600px) {
  .container .mobile-nav {
    display: none !important;
  }
}

@media screen and (max-width: 758px) {
  .close-icon {
    left: 0;
  }

  .navbar-collapse {
    ul {
      display: none !important;
    }
  }

  .sidenav-collapse {
    display: none !important;
  }

  #navbarSupportedContent .bmd-form-group {
    width: 100%;
  }

  .navbar {
    padding: 1rem 1rem !important;
    
    &-brand {
      margin-left: 0 !important;
      padding-top: 0;
    }
  }
}

.sidenav-collapse {
  width: 50px;
  height: 50px;
  line-height: 50px;
  margin-left: 10px !important;
  padding: 0 !important;
  border-radius: 50%;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;

  ion-icon {
    font-size: 25px;
  }

  &:focus {
    outline: 0;
    box-shadow: none;
    border-color: transparent;
    box-shadow: rgba(0, 0, 0, 0.1) 0px 5px 15px, rgba(0, 0, 0, 0.1) 0px 4px 30px;
  }

  &:hover {
    box-shadow: rgba(0, 0, 0, 0.1) 0px 5px 15px, rgba(0, 0, 0, 0.1) 0px 4px 30px;
    cursor: pointer;
  }
}

.user-detail {
	padding: .8rem 0 .8rem 1.5rem;
	font-weight: 600;
	color: rgba(0, 0, 0, .8);
	font-size: 14px;
}

.user-toggle {
  .avatar-placeholder {
    width: 30px;
    height: 30px;
    line-height: 30px;
    font-size: 15px;
  }
}
