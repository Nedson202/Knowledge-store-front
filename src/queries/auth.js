import { gql } from 'apollo-boost';

const addUser = gql`
  mutation($username: String!, $password: String!, $email: String!) {
    addUser(username: $username, password: $password, email: $email) {
      token
    }
  }
`;

const verifyEmail = gql`
  mutation($token: String!) {
    verifyEmail(token: $token) {
      token,
      message
    }
  }
`;

const sendVerificationEmail = gql`
  query($email: String!) {
    sendVerificationEmail(email: $email) {
      message
    }
  }
`;

const forgotPassword = gql`
  query($email: String!) {
    forgotPassword(email: $email) {
      message
    }
  }
`;

const resetPassword = gql`
  mutation($id: String!, $email: String!, $password: String!, $token: String!) {
    resetPassword(id: $id, email: $email, password: $password, token: $token) {
      message
      token
    }
  }
`;

const editProfile = gql`
  mutation($username: String, $email: String, $picture: String) {
    editProfile(username: $username, email: $email, picture: $picture) {
      token
      message
    }
  }
`;

const changePassword = gql`
  mutation($oldPassword: String!, $newPassword: String!) {
    changePassword(oldPassword: $oldPassword, newPassword: $newPassword) {
      token
      message
    }
  }
`;

const loginUser = gql`
mutation($username: String!, $password: String!) {
  loginUser(username: $username, password: $password) {
    username,
    email,
    token
  }
}
`;

const getBooks = gql`
  {
    books {
      name,
      authors,
      genre,
      year
    }
  }
`;

export {
  addUser, loginUser, getBooks, changePassword,
  verifyEmail, sendVerificationEmail,
  forgotPassword, resetPassword, editProfile
};