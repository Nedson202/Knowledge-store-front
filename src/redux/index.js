import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';

import rootReducer from './reducers/rootReducer';

const enhancer = composeWithDevTools(applyMiddleware(thunk));

const store = createStore(
  rootReducer,
  enhancer
);

export default store;
