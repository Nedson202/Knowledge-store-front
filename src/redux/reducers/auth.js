import isEmpty from 'lodash/isEmpty';
import { SET_LOGGEDIN_USER } from '../constants/actionTypes';

const initialState = {
  isAuthenticated: false,
  user: {}
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_LOGGEDIN_USER: {
      return {
        ...state,
        isAuthenticated: action.user.id && !isEmpty(action.user),
        user: action.user
      };
    }

    default: return state;
  }
};