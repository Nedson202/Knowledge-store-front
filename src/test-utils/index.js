import React from 'react';
import { MockedProvider } from 'react-apollo/test-utils';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import store from '../redux';

// eslint-disable-next-line react/prop-types
export const AllProviders = ({ children, customStore, apolloMocks }) => (
  <MockedProvider
    addTypename={false}
    mocks={apolloMocks}
  >
    <Provider store={customStore || store}>
      <BrowserRouter>
        {children}
      </BrowserRouter>
    </Provider>
  </MockedProvider>
);

export * from '@testing-library/react';

export const waitForTime = (waitFor = 0) => new Promise(resolve => setTimeout(resolve, waitFor));
