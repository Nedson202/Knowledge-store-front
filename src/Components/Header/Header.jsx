import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import avatar from '../../assets/beverage-books.jpg';
import './_Header.scss';

class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isSideNavOpen: false,
    };
  }

  toggleSideBar = () => {
    const { isSideNavOpen } = this.state;
    if (!isSideNavOpen) {
      document.addEventListener('click', this.handleOutsideClick, false);
      document.getElementById('mySidenav').style.width = '270px';
      document.getElementById('mySidenav').style.boxShadow = '0 1px 1px 100vw rgba(0, 0, 0, 0.6)';
    } else {
      document.removeEventListener('click', this.handleOutsideClick, false);
      document.getElementById('mySidenav').style.width = '0';
      document.getElementById('mySidenav').style.boxShadow = 'none';
    }

    this.setState(prevState => ({
      isSideNavOpen: !prevState.isSideNavOpen,
    }));
  }

  handleOutsideClick = (event) => {
    if (this.node.contains(event.target)) {
      return;
    }

    this.toggleSideBar();
  }

  render() {
    return (
      <div className="App" ref={(node) => { this.node = node; }}>
        <nav className="navbar fixed-top navbar-expand-lg navbar-light bg-light" id="navbar">
          <div className="container">
            <span className="navbar-brand">Loresters Bookstore</span>
            <button
              type="button"
              className="mobile-nav"
              data-toggle="collapse"
              data-target="#navbarSupportedContent"
              aria-controls="navbarSupportedContent"
              aria-expanded="false"
              aria-label="Search icon"
            >
              <i className="fas fa-search" />
            </button>
            <button
              type="button"
              className="mobile-nav"
              aria-label="Toggle navigation"
              onClick={this.toggleSideBar}
            >
              <i className="fas fa-bars" />
            </button>

            <div className="collapse navbar-collapse" id="navbarSupportedContent">
              <form className="form-inline my-2 my-lg-0 ml-auto mt-2 mt-lg-0">
                <i className="fa fa-search" />
                <input
                  className="form-control mr-sm-2 search-bar"
                  type="search"
                  placeholder="Search by book name, author, year"
                  aria-label="Search"
                />
              </form>
              <ul className="navbar-nav ml-auto mt-2 mt-lg-0">
                <li className="cart-navbar">
                  <i className="fa fa-shopping-cart" />
                  <span>10</span>
                </li>
                <li className="nav-item dropdown user-profile-nav">
                  <button
                    type="button"
                    className="btn btn-default nav-link dropdown-toggle text-case"
                    id="navbarDropdown"
                    data-toggle="dropdown"
                    aria-haspopup="true"
                    aria-expanded="false"
                  >
                    <img src={avatar} alt="Avatar" className="avatar" />
                  </button>
                  <div className="dropdown-menu" aria-labelledby="navbarDropdown">
                    <div className="user-detail">
                      @michaelSmith
                    </div>
                    <div className="dropdown-divider" />
                    <Link to="/profile" className="dropdown-item user-profile-navlink">
                      Profile
                      {' '}
                      <i className="fas fa-user-circle" />
                    </Link>
                    <Link to="/logout" className="dropdown-item user-profile-navlink">
                      Logout
                      {' '}
                      <i className="fa fa-sign-out-alt" />
                    </Link>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </nav>
      </div>
    );
  }
}

export default Header;
