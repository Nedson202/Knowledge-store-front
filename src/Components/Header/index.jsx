import React, { Component, Fragment } from 'react';
import { Link, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import './_Header.scss';
import Avatar from '../ReviewCard/Avatar';
import { logOutUser } from '../../redux/actions/userActions';
import Search from '../Search';
import Login from '../Login';
import SignUp from '../SignUp';
import {
  OPEN, CLICK, BLOCK, NONE, AUTO, LEFT_SIDE_BAR,
  MAIN, CLOSED, SIDE_BAR_TEXT, SIDE_BAR_STATUS, SIDE_NAV
} from '../../defaults';

class Header extends Component {
  state = {
    isSideNavOpen: false,
    isSideBarOpen: false,
  };

  componentDidMount() {
    return localStorage.sideBarStatus !== OPEN && this.toggleSidebar();
  }

  componentWillUnmount() {
    document.removeEventListener(CLICK, this.handleOutsideClick, false);
  }

  toggleSidebar = () => {
    const { isSideBarOpen } = this.state;
    const mainContent = document.getElementById(MAIN);

    if (isSideBarOpen) {
      localStorage.setItem(SIDE_BAR_STATUS, OPEN);
      document.getElementById(LEFT_SIDE_BAR).style.width = '260px';
      this.toggleSidebarText(BLOCK);
    } else {
      localStorage.setItem(SIDE_BAR_STATUS, CLOSED);
      document.getElementById(LEFT_SIDE_BAR).style.width = '70px';

      if (mainContent) {
        mainContent.style.marginLeft = AUTO;
      }
      this.toggleSidebarText(NONE);
    }
    this.setState(prevState => ({
      isSideBarOpen: !prevState.isSideBarOpen,
    }));
  };

  toggleSidebarText = (display) => {
    const texts = document.querySelectorAll(SIDE_BAR_TEXT);
    for (let text = 0; text < texts.length; text += 1) {
      texts[text].style.display = display;
    }
  };

  toggleMobileNav = () => {
    const { isSideNavOpen } = this.state;
    const sideNavEL = document.getElementById(SIDE_NAV);
    if (!isSideNavOpen) {
      document.addEventListener(CLICK, this.handleOutsideClick, false);
      sideNavEL.style.width = '270px';
      sideNavEL.style.boxShadow = '0 1px 1px 100vw rgba(0, 0, 0, 0.6)';
    } else {
      document.removeEventListener(CLICK, this.handleOutsideClick, false);
      sideNavEL.style.width = '0';
      sideNavEL.style.boxShadow = NONE;
    }

    this.setState(prevState => ({
      isSideNavOpen: !prevState.isSideNavOpen,
    }));
  }

  handleOutsideClick = (event) => {
    if (this.node.contains(event.target)) {
      return;
    }

    this.toggleMobileNav();
  }

  handleLogout = () => {
    const { dispatch, history } = this.props;
    dispatch(logOutUser());
    history.push('/books');
  }

  authenticationForms() {
    return (
      <Fragment>
        <Login />
        <SignUp />
      </Fragment>
    );
  }

  emailConfirmationNote() {
    return (
      <div className="pending-verification">
        A verification link has been sent to your mail box.
        Click on the link or use this
        {' '}
        <Link
          to="/email?verify-email=true"
          style={{ color: 'black' }}
        >
          Link
        </Link>
        {' '}
        to request for a new verification mail.
      </div>
    );
  }

  renderAuthButtons() {
    return (
      <Fragment>
        <button
          type="button"
          className="btn btn-default btn-raised cancel-button btn"
          data-toggle="modal"
          data-target="#LoginFormModal"
        >
          Login
        </button>
        <button
          type="button"
          className="btn btn-primary btn-raised text-case login-button"
          data-toggle="modal"
          data-target="#SignUpFormModal"
        >
          Signup
        </button>
      </Fragment>
    );
  }

  renderUserAvatar() {
    const { user: { username, picture, avatarColor, } = {} } = this.props;
    return (
      <li className="nav-item dropdown user-profile-nav">
        <button
          type="button"
          className="btn btn-default nav-link dropdown-toggle text-case user-toggle"
          id="navbarDropdown"
          data-toggle="dropdown"
          aria-haspopup="true"
          aria-expanded="false"
        >
          {picture && <img src={picture} alt="Avatar" className="avatar" />}
          {!picture && <Avatar user={username} color={avatarColor} />}
        </button>
        <div className="dropdown-menu" aria-labelledby="navbarDropdown">
          <div className="user-detail">
            {username}
          </div>
          <div className="dropdown-divider" />
          <Link to="/profile" className="dropdown-item">
            <ion-icon class="user-profile-icon" name="person" />
            {' '}
            Profile
          </Link>
          <button type="button" className="dropdown-item" onClick={this.handleLogout}>
            <ion-icon class="user-profile-icon" name="log-out" />
            {' '}
            Logout
          </button>
        </div>
      </li>
    );
  }

  render() {
    const { user } = this.props;
    return (
      <Fragment>
        <div className="App" ref={(node) => { this.node = node; }}>
          <nav
            className="navbar fixed-top navbar-expand-lg navbar-light bg-light"
            id="navbar"
          >
            <div
              onClick={this.toggleSidebar}
              className="sidenav-collapse
              dropdown-item sidebar-navlink collapse-bar btn-raised"
              data-tip="Toggle sidebar"
              role="button"
              tabIndex={0}
            >
              <ion-icon name="menu" />
            </div>
            <div className="container">
              <Link to="/">
                <span className="navbar-brand">Loresters Bookstore</span>
              </Link>
              <div>
                <button
                  type="button"
                  className="mobile-nav"
                  data-toggle="collapse"
                  data-target="#navbarSupportedContent"
                  aria-controls="navbarSupportedContent"
                  aria-expanded="false"
                  aria-label="Search icon"
                >
                  <ion-icon name="search" />
                </button>
                <button
                  type="button"
                  className="mobile-nav"
                  aria-label="Toggle navigation"
                  onClick={this.toggleMobileNav}
                >
                  <ion-icon name="menu" />
                </button>
              </div>

              <div
                className="collapse navbar-collapse"
                id="navbarSupportedContent"
              >
                <Search />
                <ul className="navbar-nav ml-auto mt-2 mt-lg-0">
                  {user.isVerified === 'true' && this.renderUserAvatar()}
                </ul>
                {user.isVerified !== 'true' && this.renderAuthButtons()}
              </div>
            </div>
          </nav>
          {this.authenticationForms()}
        </div>
      </Fragment>
    );
  }
}

const mapStateToProps = ({ auth }) => ({
  isAuthenticated: auth.isAuthenticated,
  user: auth.user
});

Header.propTypes = {
  dispatch: PropTypes.func,
  history: PropTypes.object,
  user: PropTypes.object,
};

Header.defaultProps = {
  dispatch: () => { },
  history: {},
  user: {},
};

export default withRouter(connect(mapStateToProps)(Header));
