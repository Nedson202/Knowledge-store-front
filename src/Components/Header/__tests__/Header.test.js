import React from 'react';
import { createStore } from 'redux';

import { render, fireEvent, AllProviders } from 'test-utils';
import { LOGOUT } from 'settings';

import auth from '../../../redux/reducers/auth';
import Header from '..';

describe('Header', () => {
  it('should render darkmode switch', () => {
    const {
      container, asFragment, getByText, queryAllByTestId
    } = render(
      <AllProviders>
        <Header />
      </AllProviders>
    );

    const darkModeSwitch = queryAllByTestId('dark-mode-switch');

    getByText('Loresters Bookstore');

    expect(asFragment()).toMatchSnapshot();
    expect(container).toBeInTheDocument();
    expect(darkModeSwitch[0]).toBeInTheDocument();
  });

  it('should display auth buttons if not signed in', () => {
    const mockProps = {
      auth: {
        isAuthenticated: false,
      }
    };

    const store = createStore(auth, mockProps);

    const { getByText, queryByTestId } = render(
      <AllProviders
        customStore={store}
      >
        <Header />
      </AllProviders>
    );

    getByText('Login');
    getByText('Signup');

    expect(queryByTestId('Logout')).toBeNull();
  });

  it('should display auth buttons if not signed in', () => {
    const mockProps = {
      auth: {
        user: {
          isVerified: true,
          username: 'Cooper AL',
          picture: 'https://image.url',
          avatarColor: 'green',
        },
      }
    };

    const store = createStore(auth, mockProps);

    const { getByText, queryByTestId } = render(
      <AllProviders
        customStore={store}
      >
        <Header />
      </AllProviders>
    );

    getByText('Cooper AL');
    getByText('Profile');
    getByText('Logout');

    expect(queryByTestId('Login')).toBeNull();
  });

  it('should test theme mode', () => {
    const { queryAllByTestId } = render(
      <AllProviders>
        <Header />
      </AllProviders>
    );

    const darkModeSwitch = queryAllByTestId('dark-mode-switch');

    fireEvent.click(darkModeSwitch[0]);

    const darkThemeMode = window.document.documentElement.getAttribute('data-theme');
    expect(darkThemeMode).toBe('dark');

    fireEvent.click(darkModeSwitch[0]);

    const lightThemeMode = window.document.documentElement.getAttribute('data-theme');
    expect(lightThemeMode).toBe('light');
  });

  it('should test theme toggle storage event', () => {
    render(
      <AllProviders>
        <Header />
      </AllProviders>
    );

    window.dispatchEvent(new StorageEvent('storage', {
      key: 'theme',
      newValue: 'light'
    }));

    const lightThemeMode = window.document.documentElement.getAttribute('data-theme');
    expect(lightThemeMode).toBe('light');
  });

  it('should test logout storage event', () => {
    window.location.reload = jest.fn();

    render(
      <AllProviders>
        <Header />
      </AllProviders>
    );

    window.dispatchEvent(new StorageEvent('storage', {
      key: LOGOUT,
      newValue: 'light'
    }));

    window.location.reload.mockClear();
  });
});
