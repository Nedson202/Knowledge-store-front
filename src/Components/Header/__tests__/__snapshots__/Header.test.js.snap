// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render darkmode switch 1`] = `
<DocumentFragment>
  <nav
    class="navbar fixed-top navbar-expand-lg"
    id="navbar"
  >
    <div
      class="container"
    >
      <button
        aria-label="Toggle navigation"
        class="mobile-nav"
        data-testid="mobile-menu"
        type="button"
      >
        <ion-icon
          name="menu"
        />
      </button>
      <a
        href="/"
      >
        <span
          class="navbar-brand"
        >
          Loresters Bookstore
        </span>
      </a>
      <button
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Search icon"
        class="mobile-nav"
        data-target="#navbarSupportedContent"
        data-toggle="collapse"
        type="button"
      >
        <ion-icon
          name="search"
        />
      </button>
      <button
        aria-label="Dark mode switch"
        class="dark-mode-switch mobile-nav"
        data-testid="dark-mode-switch"
        type="button"
      >
        <ion-icon
          class="dark-mode-icon"
          name="contrast"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <form
          class="search"
        >
          <ion-icon
            name="search"
          />
          <input
            aria-label="Search"
            autocomplete="off"
            class="form-input"
            id="searchBox"
            name="value"
            placeholder="Search book collections..."
            type="search"
            value=""
          />
        </form>
        <ul
          class="navbar-nav ml-auto mt-2 mt-lg-0"
        >
          <div
            class="desktop-and-tablet"
          >
            <button
              class="btn btn-default cancel-button btn"
              data-target="#LoginFormModal"
              data-toggle="modal"
              id="login-button"
              type="button"
            >
              Login
            </button>
            <button
              class="btn btn-primary btn-raised text-case login-button"
              data-target="#SignUpFormModal"
              data-toggle="modal"
              id="signup-button"
              type="button"
            >
              Signup
            </button>
          </div>
        </ul>
        <span
          class="desktop-and-tablet"
        >
          <button
            aria-label="Dark mode switch"
            class="dark-mode-switch undefined"
            data-testid="dark-mode-switch"
            type="button"
          >
            <ion-icon
              class="dark-mode-icon"
              name="contrast"
            />
          </button>
        </span>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
